File: src/main/repairs.ts
import path from "node:path";

import { getRepairAction, REPAIR_ACTIONS } from "../shared/action-catalog";
import type {
  ActionPreview,
  ActionResult,
  RepairActionDefinition,
  RepairActionId,
  RestoreResult,
  RestoreScope,
} from "../shared/contracts";
import { AuditService } from "./audit";
import { escapePowerShellSingleQuoted, runElevatedPowerShell, runPowerShell } from "./powershell";
import { getRepairScript } from "./repair-scripts";
import { SnapshotService } from "./snapshots";

const VALID_ACTION_IDS = new Set(REPAIR_ACTIONS.map((action) => action.id));
const VALID_RESTORE_SCOPES = new Set<RestoreScope>(["proxy", "dns", "hosts", "firewall"]);

const validateActionId = (value: unknown): RepairActionId => {
  if (typeof value !== "string" || !VALID_ACTION_IDS.has(value as RepairActionId)) {
    throw new Error("拒绝执行未列入白名单的操作");
  }
  return value as RepairActionId;
};

const runStandardPowerShell = async (body: string): Promise<string[]> => {
  const output = await runPowerShell(`
$logs = [System.Collections.Generic.List[string]]::new()
function Add-Log([string]$message) { $logs.Add($message) }
${body}
@($logs) | ConvertTo-Json -Compress
`);
  if (!output) return [];
  const parsed = JSON.parse(output) as string[] | string;
  return Array.isArray(parsed) ? parsed : [parsed];
};

export class RepairService {
  constructor(
    private readonly dataDirectory: string,
    private readonly snapshots: SnapshotService,
    private readonly audit: AuditService,
  ) {}

  list(): RepairActionDefinition[] {
    return REPAIR_ACTIONS;
  }

  preview(rawActionId: unknown): ActionPreview {
    const action = getRepairAction(validateActionId(rawActionId));
    const warnings: string[] = [];
    if (action.risk === "medium") warnings.push("执行期间网络可能短暂中断");
    if (action.risk === "high") warnings.push("这是高风险操作，请先确认没有依赖自定义网络配置");
    if (action.restartRequired) warnings.push("操作完成后需要重启 Windows 才能完全生效");
    if (action.id === "full-network-reset") warnings.push("虚拟网卡和 VPN/TUN 驱动也会被移除并重新安装");
    return { action, snapshotWillBeCreated: true, warnings };
  }

  async run(rawActionId: unknown): Promise<ActionResult> {
    const actionId = validateActionId(rawActionId);
    const action = getRepairAction(actionId);
    const snapshot = await this.snapshots.create(`执行前备份：${action.title}`);
    const snapshotDirectory = this.snapshots.resolveDirectory(snapshot.id);
    const script = getRepairScript(actionId, snapshotDirectory);
    try {
      const logs = action.requiresAdmin
        ? (await runElevatedPowerShell(script, path.join(this.dataDirectory, "jobs"))).logs
        : await runStandardPowerShell(script);
      const message = `${action.title}已完成${action.restartRequired ? "，请重启电脑" : ""}`;
      await this.audit.write({
        kind: "repair",
        title: action.title,
        detail: `${message} · 回滚快照 ${snapshot.id}`,
        success: true,
      });
      return {
        success: true,
        actionId,
        snapshotId: snapshot.id,
        message,
        logs,
        restartRequired: action.restartRequired,
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      await this.audit.write({
        kind: "error",
        title: `${action.title}失败`,
        detail: `${message} · 配置快照 ${snapshot.id} 已保留`,
        success: false,
      });
      return {
        success: false,
        actionId,
        snapshotId: snapshot.id,
        message,
        logs: [],
        restartRequired: false,
      };
    }
  }

  async restore(rawSnapshotId: unknown, rawScopes: unknown): Promise<RestoreResult> {
    if (typeof rawSnapshotId !== "string") throw new Error("快照编号无效");
    if (!Array.isArray(rawScopes) || rawScopes.length === 0) throw new Error("请至少选择一项还原内容");
    const scopes = [...new Set(rawScopes.map((scope) => {
      if (typeof scope !== "string" || !VALID_RESTORE_SCOPES.has(scope as RestoreScope)) {
        throw new Error("还原范围不在白名单中");
      }
      return scope as RestoreScope;
    }))];
    await this.snapshots.read(rawSnapshotId);
    const snapshotDirectory = this.snapshots.resolveDirectory(rawSnapshotId);
    const snapshotPath = escapePowerShellSingleQuoted(path.join(snapshotDirectory, "snapshot.json"));
    const hostsPath = escapePowerShellSingleQuoted(path.join(snapshotDirectory, "hosts"));
    const firewallPath = escapePowerShellSingleQuoted(path.join(snapshotDirectory, "firewall.wfw"));
    const blocks: string[] = [
      `$stored = Get-Content -LiteralPath '${snapshotPath}' -Raw | ConvertFrom-Json`,
    ];
    if (scopes.includes("proxy")) {
      blocks.push(`
$key = 'HKCU:\\Software\\Microsoft\\Windows\\CurrentVersion\\Internet Settings'
$proxy = $stored.raw.proxy
Set-ItemProperty -LiteralPath $key -Name ProxyEnable -Type DWord -Value $(if ($proxy.enabled) { 1 } else { 0 })
foreach ($pair in @(@('ProxyServer', [string]$proxy.server), @('ProxyOverride', [string]$proxy.override), @('AutoConfigURL', [string]$proxy.autoConfigUrl))) {
  if ($pair[1]) { Set-ItemProperty -LiteralPath $key -Name $pair[0] -Type String -Value $pair[1] }
  else { Remove-ItemProperty -LiteralPath $key -Name $pair[0] -ErrorAction SilentlyContinue }
}
Set-ItemProperty -LiteralPath $key -Name AutoDetect -Type DWord -Value $(if ($proxy.autoDetect) { 1 } else { 0 })
Add-Log '用户代理设置已还原'
`);
    }
    if (scopes.includes("dns")) {
      blocks.push(`
$restored = 0
foreach ($entry in @($stored.raw.dns | Where-Object { [int]$_.addressFamily -eq 2 })) {
  $adapter = Get-NetAdapter -InterfaceIndex ([int]$entry.interfaceIndex) -ErrorAction SilentlyContinue
  if (!$adapter) { continue }
  $servers = @($entry.serverAddresses | ForEach-Object { [string]$_ })
  if ($servers.Count -eq 0) { Set-DnsClientServerAddress -InterfaceIndex $adapter.ifIndex -ResetServerAddresses }
  else { Set-DnsClientServerAddress -InterfaceIndex $adapter.ifIndex -ServerAddresses $servers }
  $restored++
}
Clear-DnsClientCache -ErrorAction SilentlyContinue
Add-Log ("已还原 {0} 个网卡的 DNS" -f $restored)
`);
    }
    if (scopes.includes("hosts")) {
      blocks.push(`
if (!(Test-Path -LiteralPath '${hostsPath}')) { throw '此快照没有可用的 Hosts 备份' }
Copy-Item -LiteralPath '${hostsPath}' -Destination (Join-Path $env:SystemRoot 'System32\\drivers\\etc\\hosts') -Force
Clear-DnsClientCache -ErrorAction SilentlyContinue
Add-Log 'Hosts 文件已还原'
`);
    }
    if (scopes.includes("firewall")) {
      blocks.push(`
if (!(Test-Path -LiteralPath '${firewallPath}')) { throw '此快照没有防火墙策略文件' }
netsh advfirewall import '${firewallPath}' | Out-Null
if ($LASTEXITCODE -ne 0) { throw '防火墙策略还原失败' }
Add-Log '防火墙策略已还原'
`);
    }

    try {
      const result = await runElevatedPowerShell(blocks.join("\n"), path.join(this.dataDirectory, "jobs"));
      await this.audit.write({
        kind: "restore",
        title: "网络配置已还原",
        detail: `${rawSnapshotId} · ${scopes.join("、")}`,
        success: true,
      });
      return {
        success: true,
        snapshotId: rawSnapshotId,
        scopes,
        message: "所选网络配置已还原",
        logs: result.logs,
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      await this.audit.write({
        kind: "error",
        title: "网络配置还原失败",
        detail: `${rawSnapshotId} · ${message}`,
        success: false,
      });
      return { success: false, snapshotId: rawSnapshotId, scopes, message, logs: [] };
    }
  }
}

File: src-tauri/src/repairs.rs
use std::{collections::HashSet, path::Path, time::Duration};

use serde_json::Value;

use crate::{
    audit,
    models::{ActionResult, RestoreResult},
    powershell::{escape_single_quoted, run_elevated, run_powershell},
    snapshots,
};

const REPAIR_SCRIPT: &str = include_str!("../resources/repair.ps1");
const RESTORE_SCRIPT: &str = include_str!("../resources/restore.ps1");

struct ActionMeta {
    title: &'static str,
    requires_admin: bool,
    restart_required: bool,
}

fn action_meta(id: &str) -> Result<ActionMeta, String> {
    let meta = match id {
        "quick-repair" => ("智能安全修复", true, false),
        "flush-dns" => ("刷新 DNS 缓存", false, false),
        "renew-dhcp" => ("重新获取 IP", true, false),
        "dns-auto" => ("DNS 恢复自动", true, false),
        "disable-dead-proxy" => ("关闭失效系统代理", false, false),
        "reset-winhttp-proxy" => ("重置 WinHTTP 代理", true, false),
        "sync-winhttp-proxy" => ("同步系统代理到 WinHTTP", true, false),
        "normalize-proxy-bypass" => ("补全局域网直连规则", false, false),
        "remove-orphan-tun-routes" => ("清理断开的 TUN 路由", true, false),
        "restart-active-adapters" => ("重启当前网卡", true, false),
        "reset-winsock" => ("重置 Winsock", true, true),
        "reset-tcpip" => ("重置 TCP/IP", true, true),
        "reset-firewall" => ("重置防火墙规则", true, false),
        "reset-hosts" => ("重置 Hosts 文件", true, false),
        "full-network-reset" => ("彻底重装网络组件", true, true),
        _ => return Err("拒绝执行未列入白名单的操作".to_string()),
    };
    Ok(ActionMeta {
        title: meta.0,
        requires_admin: meta.1,
        restart_required: meta.2,
    })
}

fn parse_logs(output: &str) -> Vec<String> {
    if output.is_empty() {
        return Vec::new();
    }
    match serde_json::from_str::<Value>(output) {
        Ok(Value::Array(items)) => items
            .into_iter()
            .filter_map(|item| item.as_str().map(str::to_string))
            .collect(),
        Ok(Value::String(item)) => vec![item],
        _ => Vec::new(),
    }
}

fn run_standard(data_dir: &Path, body: &str) -> Result<Vec<String>, String> {
    let script = format!(
        r#"
$logs = [System.Collections.Generic.List[string]]::new()
function Add-Log([string]$message) {{ $logs.Add($message) }}
{body}
ConvertTo-Json -InputObject @($logs) -Compress
"#
    );
    run_powershell(data_dir, &script, Duration::from_secs(180)).map(|output| parse_logs(&output))
}

pub fn run(data_dir: &Path, action_id: &str) -> Result<ActionResult, String> {
    let meta = action_meta(action_id)?;
    let snapshot = snapshots::create(data_dir, Some(&format!("执行前备份：{}", meta.title)))?;
    let snapshot_dir = snapshots::resolve(data_dir, &snapshot.id)?;
    let script = REPAIR_SCRIPT
        .replace("__ACTION_ID__", &escape_single_quoted(action_id))
        .replace(
            "__SNAPSHOT_DIRECTORY__",
            &escape_single_quoted(&snapshot_dir.to_string_lossy()),
        );
    let outcome = if meta.requires_admin {
        run_elevated(data_dir, &script, Duration::from_secs(240))
    } else {
        run_standard(data_dir, &script)
    };
    match outcome {
        Ok(logs) => {
            let message = format!(
                "{}已完成{}",
                meta.title,
                if meta.restart_required {
                    "，请重启电脑"
                } else {
                    ""
                }
            );
            let _ = audit::write(
                data_dir,
                "repair",
                meta.title,
                &format!("{message} · 回滚快照 {}", snapshot.id),
                true,
            );
            Ok(ActionResult {
                success: true,
                action_id: action_id.to_string(),
                snapshot_id: Some(snapshot.id),
                message,
                logs,
                restart_required: meta.restart_required,
            })
        }
        Err(message) => {
            let _ = audit::write(
                data_dir,
                "error",
                &format!("{}失败", meta.title),
                &format!("{message} · 配置快照 {} 已保留", snapshot.id),
                false,
            );
            Ok(ActionResult {
                success: false,
                action_id: action_id.to_string(),
                snapshot_id: Some(snapshot.id),
                message,
                logs: Vec::new(),
                restart_required: false,
            })
        }
    }
}

pub fn restore(
    data_dir: &Path,
    snapshot_id: &str,
    raw_scopes: Vec<String>,
) -> Result<RestoreResult, String> {
    if raw_scopes.is_empty() {
        return Err("请至少选择一项还原内容".to_string());
    }
    let allowed: HashSet<&str> = ["proxy", "dns", "hosts", "firewall"].into_iter().collect();
    let mut seen = HashSet::new();
    let scopes: Vec<String> = raw_scopes
        .into_iter()
        .filter(|scope| seen.insert(scope.clone()))
        .map(|scope| {
            if allowed.contains(scope.as_str()) {
                Ok(scope)
            } else {
                Err("还原范围不在白名单中".to_string())
            }
        })
        .collect::<Result<_, _>>()?;
    snapshots::read(data_dir, snapshot_id)?;
    let directory = snapshots::resolve(data_dir, snapshot_id)?;
    let scope_literal = scopes
        .iter()
        .map(|scope| format!("'{}'", escape_single_quoted(scope)))
        .collect::<Vec<_>>()
        .join(",");
    let script = RESTORE_SCRIPT
        .replace(
            "__SNAPSHOT_PATH__",
            &escape_single_quoted(&directory.join("snapshot.json").to_string_lossy()),
        )
        .replace(
            "__HOSTS_PATH__",
            &escape_single_quoted(&directory.join("hosts").to_string_lossy()),
        )
        .replace(
            "__FIREWALL_PATH__",
            &escape_single_quoted(&directory.join("firewall.wfw").to_string_lossy()),
        )
        .replace("__SCOPES__", &scope_literal);
    match run_elevated(data_dir, &script, Duration::from_secs(240)) {
        Ok(logs) => {
            let _ = audit::write(
                data_dir,
                "restore",
                "网络配置已还原",
                &format!("{snapshot_id} · {}", scopes.join("、")),
                true,
            );
            Ok(RestoreResult {
                success: true,
                snapshot_id: snapshot_id.to_string(),
                scopes,
                message: "所选网络配置已还原".to_string(),
                logs,
            })
        }
        Err(message) => {
            let _ = audit::write(
                data_dir,
                "error",
                "网络配置还原失败",
                &format!("{snapshot_id} · {message}"),
                false,
            );
            Ok(RestoreResult {
                success: false,
                snapshot_id: snapshot_id.to_string(),
                scopes,
                message,
                logs: Vec::new(),
            })
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn only_known_actions_are_allowed() {
        assert!(action_meta("quick-repair").is_ok());
        assert!(action_meta("run-anything").is_err());
    }
}
